src/PotentialSimulation/ConvergenceAndRefinement.jl

```diff
@@ -88,282 +88,181 @@ function _update_till_convergence!( fssrb::PotentialSimulationSetupRB{T, N1, N2}
-            while b
-                if length(diffs) > 0
-                    idx = findmax(diffs)[2]
-                    if !in(idx, inds_φ)
-                        append!(inds_φ, idx)
-                        b = false
-                    else
-                        deleteat!(diffs, idx)
-                    end
-                else
-                    for i in eachindex(axφ)
-                        if !in(i, inds_φ) push!(inds_φ, i) end
-                    end
-                    b = false
-                end
-            end
-        end    
-        @assert iseven(length(inds_φ)) "Refinement would result in uneven grid in φ."
     end
+    return _convert_closed_potential(typeof(p), ElectricPotential(new_data, new_grid))
+end             
 
-    if isodd(length(inds_z))
-        for iz in 1:size(potential, 3)
-            if !in(iz, inds_z)
-                if iz != size(potential, 3)
-                    if axz[iz + 1] - axz[iz] > minimum_distance_z
-                        append!(inds_z, iz)
-                        break
-                    end
-                end
-            end
-        end
-    end
-    if isodd(length(inds_z)) && ((length(inds_z) + 1) < length(axz))
-        b = true
-        diffs = diff(axz)
-        while b
-            if length(diffs) > 0
-                idx = findmax(diffs)[2]
-                if !in(idx, inds_z)
-                    append!(inds_z, idx)
-                    b = false
-                else
-                    deleteat!(diffs, idx)
-                end
-            else
-                for i in eachindex(axz[1:end-1])
-                    if !in(i, inds_z) push!(inds_z, i) end
-                end
-                b = false
-            end
-        end
-    end
-    if isodd(length(inds_z)) && (length(inds_z) > 1)
-        b = true
-        diffs = diff(axz)
-        while b && length(diffs) > 1
-            idx = findmin(diffs)[2]
-            if in(idx, inds_z)
-                deleteat!(inds_z, findfirst(i -> i == idx, inds_z))
-                b = false
-            else
-                deleteat!(diffs, idx)
-            end
-        end
-    end
-    if isodd(length(inds_z)) inds_z = inds_z[1:end-1] end
-    @assert iseven(length(inds_z)) "Refinement would result in uneven grid in z."
-
-    return sort!(inds_r), sort!(inds_φ), sort!(inds_z)
+function interpolate_closed_potential(p::ScalarPotential, ::Val{true}) where {T}
+    interpolate((p.grid.axes[1], p.grid.axes[3]), p.data[:,1,:], Gridded(Linear()))
+end
+function interpolate_closed_potential(p::ScalarPotential, ::Val{false}) where {T}
+    interpolate(p.grid.axes, p.data, Gridded(Linear()))
 end
 
+_get_closed_ticks(ticks::Vector{T}, int::ClosedInterval{T}) where {T} = ticks
+_get_closed_ticks(ticks::Vector{T}, int::Interval{:closed, :open, T}) where {T} = vcat(ticks, [int.right])
+_get_closed_ticks(ticks::Vector{T}, int::Interval{:open, :closed, T}) where {T} = vcat([int.left], ticks)

```

Comment:
To reduce allocations: `vcat(ticks, int.right)` or `vcat(int.left, ticks)` without putting them in `Array` brackets